Objects/tupleobject.c

```diff
@@ -333,39 +333,42 @@ tuplerepr(PyTupleObject *v)
     return NULL;
 }
 
-/* The addend 82520, was selected from the range(0, 1000000) for
-   generating the greatest number of prime multipliers for tuples
-   up to length eight:
 
-     1082527, 1165049, 1082531, 1165057, 1247581, 1330103, 1082533,
-     1330111, 1412633, 1165069, 1247599, 1495177, 1577699
+/* Hash for tuples. This is a slightly simplified version of the xxHash
+   non-cryptographic hash:
+   - we do not use any parallellism, there is only 1 accumulator.
+   - we drop the final mixing since this is just a permutation of the
+     output space: it does not help against collisions.
+   For the xxHash specification, see
+   https://github.com/Cyan4973/xxHash/wiki/xxHash-specification-(draft)
 
    Tests have shown that it's not worth to cache the hash value, see
-   issue #9685.
+   https://bugs.python.org/issue9685
 */
-
 static Py_hash_t
 tuplehash(PyTupleObject *v)
 {
-    Py_uhash_t x;  /* Unsigned for defined overflow behavior. */
-    Py_hash_t y;
-    Py_ssize_t len = Py_SIZE(v);
-    PyObject **p;
-    Py_uhash_t mult = _PyHASH_MULTIPLIER;
-    x = 0x345678UL;
-    p = v->ob_item;
-    while (--len >= 0) {
-        y = PyObject_Hash(*p++);
-        if (y == -1)
+    Py_ssize_t i, len = Py_SIZE(v);
+    PyObject **item = v->ob_item;
+
+    Py_uhash_t acc = _PyHASH_XXPRIME_5;
+    for (i = 0; i < len; i++) {
+        Py_uhash_t lane = PyObject_Hash(item[i]);
+        if (lane == (Py_uhash_t)-1) {
             return -1;
-        x = (x ^ y) * mult;
-        /* the cast might truncate len; that doesn't change hash stability */
-        mult += (Py_hash_t)(82520UL + len + len);
+        }
+        acc += lane * _PyHASH_XXPRIME_2;
+        acc = _PyHASH_XXROTATE(acc);
+        acc *= _PyHASH_XXPRIME_1;
+    }
+
+    /* Add input length */
+    acc += len;
```

Comment:
Please add in a final permutation constant as well.  It only costs one clock cycle and may have benefit for nested tuples. Note, the original xxhash had a complex final permutation, but I think a single large random addend will suffice.

Reply:
I'll add a final permutation not because it will help with anything but just to get this PR over with.

Reply:
I changed the length addition to
```
acc += len ^ (_PyHASH_XXPRIME_5 ^ 3527539);
```
I hope that this counts as "final permutation". Since this is folded in the length, this might be outside of the critical path.

Reply:
I see no point to this (for reasons explained elsewhere - _this_ function scrambles its inputs (via multiplication) before folding them into its state, so when one of its inputs is a nested tuple's hash the latter is already getting scrambled in a more disruptive way than adding a constant), but won't fight to the death over it either.

In any case, I agree this way of writing it gives a compiler the _opportunity_ to evaluate the entire right-hand side any time after `len` is read up.   The line is utterly mysterious, though, and I'm at a loss to come with a comment that could make a case for why it's there.

Reply:
FYI, changing to

    acc += len ^ (_PyHASH_XXPRIME_5 ^ 3527539);

made no difference in about 80% of my tests (neither number of collisions nor maximum pileup changed).  The rest saw minor changes, in both directions.

The best:  the # of collisions in the 32-bit new tuple test dropped from 20 to 9.  Either is fine.

The worst:  the # of collisions taking the lower 32 bits of the 64-bit hash code in the old tuple test leapt from 0 to 7 (and was the only change in maximum pileup).  Either is fine.

Overall:  doesn't really hurt results, doesn't really help results.